feat: introduce group policies

flavio opened the pull request:
Allow multiple Kubewarden policies to be grouped together using an evaluation expression.

This PR poses the bases required to implement https://github.com/kubewarden/rfc/blob/main/rfc/0020-policy-group.md

Part of https://github.com/kubewarden/kubewarden-controller/issues/798

fabriziosestito commented on src/config.rs:
```diff
@@ -245,31 +249,112 @@ impl From<PolicyMode> for String {
+}
+
+#[derive(Debug, Clone)]
+pub enum PolicySettings {
```
I think we should use something more explicit, such as `PolicyOrPolicyGroup`

fabriziosestito commented on src/config.rs:
```diff
@@ -245,31 +249,112 @@ impl From<PolicyMode> for String {
+
+#[derive(Debug, Clone)]
+pub enum PolicySettings {
+    IndividualPolicy(SettingsJSON),
```
This could be `Policy(...)`

fabriziosestito commented on src/config.rs:
```diff
@@ -245,31 +249,112 @@ impl From<PolicyMode> for String {
+#[derive(Debug, Clone)]
+pub enum PolicySettings {
+    IndividualPolicy(SettingsJSON),
+    GroupPolicy {
```
`PolicyGroup`

fabriziosestito commented on src/config.rs:
```diff
@@ -245,31 +249,112 @@ impl From<PolicyMode> for String {
 #[derive(Deserialize, Debug, Clone)]
 #[serde(rename_all = "camelCase")]
-pub struct Policy {
+pub struct GroupPolicyMember {
```
`PolicyGroupMember`

fabriziosestito commented on src/config.rs:
```diff
@@ -245,31 +249,112 @@ impl From<PolicyMode> for String {
+    GroupPolicy {
+        expression: String,
+        message: String,
+        sub_policies: Vec<String>,
```
members

fabriziosestito commented on src/config.rs:
```diff
@@ -245,31 +249,112 @@ impl From<PolicyMode> for String {
+/// Describes a policy that can be either an individual policy or a group policy.
+#[derive(Deserialize, Debug, Clone)]
+#[serde(untagged, rename_all = "camelCase")]
+pub enum Policy {
```
`PolicyOrPolicyGroup`

fabriziosestito commented on src/config.rs:
```diff
@@ -245,31 +249,112 @@ impl From<PolicyMode> for String {
+#[serde(untagged, rename_all = "camelCase")]
+pub enum Policy {
+    /// An individual policy
+    Individual {
```
Policy

fabriziosestito commented on src/config.rs:
```diff
@@ -245,31 +249,112 @@ impl From<PolicyMode> for String {
+        context_aware_resources: BTreeSet<ContextAwareResource>,
+    },
+    /// A group of policies that are evaluated together using a given expression
+    Group {
```
PolicyGroup

fabriziosestito commented on src/config.rs:
```diff
@@ -245,31 +249,112 @@ impl From<PolicyMode> for String {
+        policy_mode: PolicyMode,
+        /// The policies that make up for this group
+        /// Key is a unique identifier
+        policies: HashMap<String, GroupPolicyMember>,
```
maybe members for consistency

fabriziosestito commented on src/evaluation/evaluation_environment.rs:
```diff
@@ -66,6 +73,10 @@ pub(crate) struct EvaluationEnvironment {
+
+    /// A Set containing the IDs of the group policies. The ID is the name of the policy as
+    /// declared inside of the `policies.yml` file.
+    group_policies: HashSet<String>,
```
policy_groups

fabriziosestito commented on src/evaluation/evaluation_environment.rs:
```diff
@@ -175,38 +271,40 @@ impl EvaluationEnvironment {
         Ok(())
     }
 
+    fn register_group_policy(
```
register_policy_group

fabriziosestito commented on src/evaluation/evaluation_environment.rs:
```diff
@@ -291,10 +399,119 @@ impl EvaluationEnvironment {
+    /// Note, this is a static method because it is invoked from within a closure that is
+    /// passed to the Rhai engine. That requires the `evaluation_environment` to be cloned,
+    /// hence to be placed inside of an `Arc`.
+    fn validate_individual_policy(
```
validate_policy

fabriziosestito commented on src/evaluation/evaluation_environment.rs:
```diff
@@ -291,10 +399,119 @@ impl EvaluationEnvironment {
+        }
+    }
+
+    /// Validate a single policy.
```
Validate a policy

fabriziosestito commented on src/evaluation/evaluation_environment.rs:
```diff
@@ -291,10 +399,119 @@ impl EvaluationEnvironment {
+        Ok(evaluator.validate(req.clone(), &settings))
+    }
+
+    fn validate_group_policy(
```
validate_policy_group

fabriziosestito replied:
docstring is missing

fabriziosestito commented on src/evaluation/errors.rs:
```diff
@@ -16,6 +16,6 @@ pub enum EvaluationError {
-    #[error("{0}")]
-    InternalError(String),
+    #[error("Group policy used as individual policy: {0}")]
+    GroupPolicyUsedAsIndividualPolicy(String),
```
This error is only returned if someone tries to rehydrate a policy group.
Could we be more explicit?

```rust
    #[error("Cannot rehydrate policy group: {0}")]
    CannotRehydratePolicyGroup(String),
```
Where the `String` is the policy_id and not a message

fabriziosestito commented on src/evaluation/evaluation_environment.rs:
```diff
@@ -224,55 +322,65 @@ impl EvaluationEnvironment {
     fn rehydrate(&self, policy_id: &str) -> Result<PolicyEvaluator> {
+        if self.group_policies.contains(policy_id) {
+            return Err(EvaluationError::GroupPolicyUsedAsIndividualPolicy(format!(
+                "attempted to rehydrate {policy_id}"
```
See: https://github.com/kubewarden/policy-server/pull/814/files#r1670197444

flavio commented:
@fabriziosestito thanks for the review, I've addressed all the feedback you gave

fabriziosestito commented on src/evaluation.rs:
```diff
@@ -3,6 +3,6 @@ mod evaluation_environment;
 pub(crate) mod precompiled_policy;
 
-// This is required to mock the `EvaluationEnvironment` inside of our tests
+// This is required to mock the `EvaluationEnvironmento` inside of our tests
```
`EvaluationEnvironment` (my bad )

fabriziosestito reviewed (approved):
LGTM, just one typo to fix.

Also, the latest commit is not a conventional commit.

flavio commented:
@fabriziosestito I've addressed your comment. As for the commit format, I plan to squash all of them into a single commit. I'm not doing semantic git commits right now

flavio commented:
I think the current code is doing pretty good. I've implemented all the missing TODOs except for these ones:

* When validating the policy group expression, ensure the returned value is a boolean. The check is not working right now, see the commented unit test
* Validate the `policy.yml` file: error when we find something that is not covered by our spec

jvanz commented on src/config.rs:
```diff
@@ -306,170 +391,122 @@ fn convert_yaml_map_to_json(
+    //         let policy = policies.get("example").unwrap();
+    //         let json_data = convert_yaml_map_to_json(serde_yaml::Mapping::from_iter(
+    //             policy
+    //                 .settings
+    //                 .as_ref()
+    //                 .unwrap()
+    //                 .iter()
+    //                 .map(|(key, value)| (serde_yaml::Value::String(key.clone()), value.clone())),
+    //         ));
+    //         assert!(json_data.is_ok());
+
+    //         let settings = json_data.unwrap();
+    //         assert_eq!(settings.get("valid_namespace").unwrap(), "valid");
+    //     }
+
+    //     #[test]
+    //     fn handle_yaml_map_with_no_data() {
+    //         let input = r#"
+    // ---
+    // example:
+    //   url: file:///tmp/namespace-validate-policy.wasm
+    //   settings: {}
+    // "#;
+    //         let policies: HashMap<String, Policy> = serde_yaml::from_str(input).unwrap();
+    //         assert!(!policies.is_empty());
+
+    //         let policy = policies.get("example").unwrap();
+    //         let json_data = convert_yaml_map_to_json(serde_yaml::Mapping::from_iter(
+    //             policy
+    //                 .settings
+    //                 .as_ref()
+    //                 .unwrap()
+    //                 .iter()
+    //                 .map(|(key, value)| (serde_yaml::Value::String(key.clone()), value.clone())),
+    //         ));
+    //         assert!(json_data.is_ok());
+
+    //         let settings = json_data.unwrap();
+    //         assert!(settings.is_empty());
+    //     }
```
Cannot we remove this comments ?

flavio replied:
ouch, good catch! Yes, let's remove these tests since they are not useful

flavio commented:
Another update:

>    Validate policies configuration file
>    
>    * Be strict during deserialization: error when something unknown is found
>    * Validate policy names: ensure they do not contain the `/` symbol

fabriziosestito reviewed (approved):
:shipit:

flavio commented:
This is really the last change I'm going to make to this PR. With 89891ea9dd87013120fb8b5f835afafb5bc8ba11 I've introduced a dedicated Rust type to identify policies. We're no longer using strings.

fabriziosestito reviewed (approved):
LGTM @flavio love the latest commit

jvanz commented:
:shipit:

viccuad commented on src/evaluation/evaluation_environment.rs:
```diff
@@ -291,10 +466,154 @@ impl EvaluationEnvironment {
+        // We create a RAW engine, which has a really limited set of built-ins available
+        let mut rhai_engine = rhai::Engine::new_raw();
+
+        // Keep track of all the evalution results of the member policies
```
```suggestion
        // Keep track of all the evaluation results of the member policies
```

flavio commented:
Last round of cleanups:

* Rename `PolicyId` to `PolicyID`
* Config: policy group policies are now stored inside of a key named `policies` instead of `members`
* `EvaluationEnvironment`: introduce builder patter, remove duplication
  between policies and group policies

viccuad reviewed (approved):
LGTM :rocket:

flavio commented on src/api/service.rs:
```diff
@@ -105,7 +107,7 @@ pub(crate) fn evaluate(
             }
             let policy_evaluation_metric = metrics::PolicyEvaluation {
-                policy_name,
+                policy_name: policy_id.to_string(),
```
We have to use `to_string()` here because `policy_name` is a `String`, while here is a `PolicyID` type.

We call `to_string()` to convert `PolicyID` using the implementation of the `Display` trait (which we provided)
